pkg/api/types.go

```diff
@@ -1546,6 +1564,119 @@ func (p *Properties) GetCustomCloudIdentitySystem() string {
 	return AzureADIdentitySystem
 }
 
+// SetCustomCloudProfileEnvironment retrives the endpoints from Azure Stack metadata endpoint and set the values for azure.Environment
+func (cs *ContainerService) SetCustomCloudProfileEnvironment() {
+	p := cs.Properties
+	if p.IsAzureStackCloud() {
+		env := p.CustomCloudProfile.Environment
+		if env.ServiceManagementEndpoint == "" || env.ActiveDirectoryEndpoint == "" || env.GraphEndpoint == "" || env.ResourceManagerVMDNSSuffix == "" {
+			metadataURL := fmt.Sprintf("%s/metadata/endpoints?api-version=1.0", strings.TrimSuffix(env.ResourceManagerEndpoint, "/"))
+
+			// Retrieve the metadata
+			httpClient := &http.Client{
+				Timeout: 15 * time.Second,
+			}
+			endpointsresp, err := httpClient.Get(metadataURL)
+			if err != nil || endpointsresp.StatusCode != 200 {
+				log.Fatalf("%s . apimodel invalid: failed to retrieve Azure Stack endpoints from %s", err, metadataURL)
+			}
+
+			body, err := ioutil.ReadAll(endpointsresp.Body)
+			if err != nil {
+				log.Fatalf("%s . apimodel invalid: failed to read the response from %s", err, metadataURL)
+			}
+
+			endpoints := AzureStackMetadataEndpoints{}
+			err = json.Unmarshal(body, &endpoints)
+			if err != nil {
+				log.Fatalf("%s . apimodel invalid: failed to parse the response from %s", err, metadataURL)
+			}
+
+			if endpoints.GraphEndpoint == "" || endpoints.Authentication == nil || endpoints.Authentication.LoginEndpoint == "" || len(endpoints.Authentication.Audiences) == 0 || endpoints.Authentication.Audiences[0] == "" {
+				log.Fatalf("%s . apimodel invalid: invalid response from %s", err, metadataURL)
+			}
+
+			env.GraphEndpoint = endpoints.GraphEndpoint
+			env.ServiceManagementEndpoint = endpoints.Authentication.Audiences[0]
+			env.GalleryEndpoint = endpoints.GalleryEndpoint
+			env.ActiveDirectoryEndpoint = endpoints.Authentication.LoginEndpoint
+			if p.CustomCloudProfile.IdentitySystem == ADFSIdentitySystem {
+				env.ActiveDirectoryEndpoint = strings.TrimSuffix(env.ActiveDirectoryEndpoint, "/")
+				env.ActiveDirectoryEndpoint = strings.TrimSuffix(env.ActiveDirectoryEndpoint, "adfs")
+			}
+
+			env.ManagementPortalURL = endpoints.PortalEndpoint
+			azsFQDNSuffix := strings.Replace(env.ResourceManagerEndpoint, fmt.Sprintf("https://management.%s.", cs.Location), "", -1)
+			azsFQDNSuffix = strings.TrimSuffix(azsFQDNSuffix, "/")
+			env.ResourceManagerVMDNSSuffix = fmt.Sprintf("cloudapp.%s", azsFQDNSuffix)
+			env.StorageEndpointSuffix = fmt.Sprintf("%s.%s", cs.Location, azsFQDNSuffix)
+			env.KeyVaultDNSSuffix = fmt.Sprintf("vault.%s.%s", cs.Location, azsFQDNSuffix)
+		}
+	}
+}
+
+// SetCustomAzureCloudSpec sets the cloud spec for Azure Stack .
+func (p *Properties) SetCustomAzureCloudSpec() {
```

Comment:
this doesn't seem like it belongs in types.go... Is it setting defaults? If so it should be in defaults.go

Reply:
or even better in defaults-custom-cloud-profile.go since it's Azure Stack specific